modules/reactor/src/meshgenerators/PinMeshGenerator.C

```diff
@@ -475,37 +430,106 @@ PinMeshGenerator::PinMeshGenerator(const InputParameters & parameters)
+  std::map<subdomain_id_type, std::vector<std::vector<subdomain_id_type>>> region_id_map{
+      {_pin_type, _region_ids}};
+  std::map<subdomain_id_type, std::vector<std::vector<std::string>>> block_name_map;
+  block_name_map[_pin_type] = _block_names;
 
-  // Update metadata at this point since values for these metadata only get set by PCCMG
-  // at generate() stage
-  if (hasMeshProperty<Real>("max_radius_meta", name() + "_2D"))
+  // Declare mesh properties that need to be moved up to the assembly level
+  if (_is_assembly)
+  {
+    declareMeshProperty(RGMB::assembly_type, _pin_type);
+    std::map<subdomain_id_type, std::vector<std::vector<subdomain_id_type>>> pin_region_id_map;
+    pin_region_id_map.insert(
+        std::pair<subdomain_id_type, std::vector<std::vector<subdomain_id_type>>>(
+            region_id_map.begin()->first, region_id_map.begin()->second));
+    declareMeshProperty(RGMB::pin_region_id_map, pin_region_id_map);
+    std::map<subdomain_id_type, std::vector<std::vector<std::string>>> pin_block_name_map;
+    pin_block_name_map.insert(std::pair<subdomain_id_type, std::vector<std::vector<std::string>>>(
+        block_name_map.begin()->first, block_name_map.begin()->second));
+    declareMeshProperty(RGMB::pin_block_name_map, pin_block_name_map);
+    declareMeshProperty(RGMB::background_block_name, std::vector<std::string>());
+    declareMeshProperty(RGMB::duct_block_names, std::vector<std::vector<std::string>>());
+  }
+  // Declare mesh properties that are only relevant to pin meshes
+  else
   {
-    const auto max_radius_meta = getMeshProperty<Real>("max_radius_meta", name() + "_2D");
-    setMeshProperty("max_radius_meta", max_radius_meta);
+    declareMeshProperty(RGMB::pin_type, _pin_type);
+    declareMeshProperty(RGMB::pin_region_ids, region_id_map);
+    declareMeshProperty(RGMB::pin_block_names, block_name_map);
   }
-  if (hasMeshProperty<unsigned int>("background_intervals_meta", name() + "_2D"))
+
+  // Set metadata to describe pin attributes
+  declareMeshProperty(RGMB::pitch, _pitch);
+  declareMeshProperty(RGMB::is_homogenized, _homogenized);
+  declareMeshProperty(RGMB::is_single_pin, _is_assembly);
+  declareMeshProperty(RGMB::ring_radii, _ring_radii);
+  declareMeshProperty(RGMB::duct_halfpitches, _duct_halfpitch);
+  declareMeshProperty(RGMB::extruded, _extrude && _mesh_dimensions == 3);
+
+  unsigned int n_axial_levels =
+      (_mesh_dimensions == 3)
+          ? getReactorParam<std::vector<unsigned int>>(RGMB::axial_mesh_intervals).size()
+          : 1;
+  std::vector<std::vector<subdomain_id_type>> ring_region_ids(
+      n_axial_levels, std::vector<subdomain_id_type>(_ring_radii.size()));
+  std::vector<std::vector<subdomain_id_type>> duct_region_ids(
+      n_axial_levels, std::vector<subdomain_id_type>(_duct_halfpitch.size()));
+  std::vector<subdomain_id_type> background_region_ids(n_axial_levels);
+
+  for (unsigned int axial_idx = 0; axial_idx < n_axial_levels; ++axial_idx)
```

Comment:
```suggestion
  for (const auto axial_idx : make_range(n_axial_levels))
```
I havent been pushing for it but we should. We try to use range-based (with make_range and index_range for vectors) as much as possible

Reply:
Changed to range-based for loops, let me know if I missed anything